src/graphql/data/Token.ts

```diff
@@ -0,0 +1,353 @@
+
+export enum TimePeriod {
+  HOUR,
+  DAY,
+  WEEK,
+  MONTH,
+  YEAR,
+  ALL,
+}
+
+function toHistoryDuration(timePeriod: TimePeriod): HistoryDuration {
+  switch (timePeriod) {
+    case TimePeriod.HOUR:
+      return 'HOUR'
+    case TimePeriod.DAY:
+      return 'DAY'
+    case TimePeriod.WEEK:
+      return 'WEEK'
+    case TimePeriod.MONTH:
+      return 'MONTH'
+    case TimePeriod.YEAR:
+      return 'YEAR'
+    case TimePeriod.ALL:
+      return 'MAX'
+  }
+}
+
+export type PricePoint = { value: number; timestamp: number }
+
+const topTokensQuery = graphql`
+  query TokenTopQuery($page: Int!, $duration: HistoryDuration!) {
+    topTokenProjects(orderBy: MARKET_CAP, pageSize: 20, currency: USD, page: $page) {
+      description
+      homepageUrl
+      twitterName
+      name
+      tokens {
+        chain
+        address
+        symbol
+      }
+      prices: markets(currencies: [USD]) {
+        ...TokenPrices
+      }
+      markets(currencies: [USD]) {
+        price {
+          value
+          currency
+        }
+        marketCap {
+          value
+          currency
+        }
+        fullyDilutedMarketCap {
+          value
+          currency
+        }
+        volume1D: volume(duration: DAY) {
+          value
+          currency
+        }
+        volume1W: volume(duration: WEEK) {
+          value
+          currency
+        }
+        volume1M: volume(duration: MONTH) {
+          value
+          currency
+        }
+        volume1Y: volume(duration: YEAR) {
+          value
+          currency
+        }
+        pricePercentChange24h {
+          currency
+          value
+        }
+        pricePercentChange1W: pricePercentChange(duration: WEEK) {
+          currency
+          value
+        }
+        pricePercentChange1M: pricePercentChange(duration: MONTH) {
+          currency
+          value
+        }
+        pricePercentChange1Y: pricePercentChange(duration: YEAR) {
+          currency
+          value
+        }
+        priceHigh52W: priceHighLow(duration: YEAR, highLow: HIGH) {
+          value
+          currency
+        }
+        priceLow52W: priceHighLow(duration: YEAR, highLow: LOW) {
+          value
+          currency
+        }
+      }
+    }
+  }
+`
+const tokenPricesFragment = graphql`
+  fragment TokenPrices on TokenProjectMarket {
+    priceHistory(duration: $duration) {
+      timestamp
+      value
+    }
+  }
+`
+
+let cachedTopTokens: TokenTopQuery$data | null
+export function useTopTokenQuery(page: number, timePeriod: TimePeriod) {
+  const topTokens = useLazyLoadQuery<TokenTopQuery>(topTokensQuery, { page, duration: toHistoryDuration(timePeriod) })
+
+  cachedTopTokens = topTokens
+
+  return topTokens
+}
+
+type CachedTopToken = NonNullable<NonNullable<TokenTopQuery$data>['topTokenProjects']>[number]
+export function checkCachedTopToken(address: string): CachedTopToken {
+  let cachedTokenData: CachedTopToken = null
+  if (cachedTopTokens) {
```

Comment:
why do this processing every time? why not forEach it once and then store the result?